kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -176,20 +205,25 @@ private static StructType convertToPhysicalSchema(
                     logicalType,
                     physicalField.getDataType(),
                     includeFieldId);
-            String physicalName = (String) physicalField
+            String physicalName = physicalField
                 .getMetadata()
-                .get(COLUMN_MAPPING_PHYSICAL_NAME_KEY);
+                .getString(COLUMN_MAPPING_PHYSICAL_NAME_KEY);
 
             if (includeFieldId) {
-                Long fieldId = (Long) physicalField
-                    .getMetadata()
-                    .get(COLUMN_MAPPING_ID_KEY);
-                FieldMetadata fieldMetadata = FieldMetadata.builder()
-                    .putLong(PARQUET_FIELD_ID_KEY, fieldId)
-                    .build();
+                Long fieldId = physicalField.getMetadata().getLong(COLUMN_MAPPING_ID_KEY);
+                FieldMetadata.Builder builder = FieldMetadata.builder()
+                    .putLong(PARQUET_FIELD_ID_KEY, fieldId);
+
+                // Nested field IDs for the 'element' and 'key'/'value' fields of Arrays/Maps
+                // are written when Uniform with IcebergCompatV2 is enabled on a table
```

Comment:
shouldn't we check if IcebergCompatV2 is enabled?

Reply:
it appears that https://github.com/delta-io/delta/blob/669dca9c05f1ddb32ce1fd612ffd83eabb1cddd9/spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L209-L213 also doesn't check that flag and assumes that the flag was true if nested column ids exist

Reply:
done